core/src/main/scala/pureconfig/ConfigSource.scala

```diff
@@ -0,0 +1,309 @@
+ */
+final class ConfigObjectSource private (config: => Result[Config]) extends ConfigSource {
+
+  def value(): Result[ConfigObject] =
+    config.right.map(_.resolve.root)
+
+  /**
+   * Merges this source with another one, with the latter being used as a fallback (e.g. the
+   * source on which this method is called takes priority). Both sources are required to produce
+   * a config object successfully.
+   *
+   * @param cs the config source to use as fallback
+   * @return a new `ConfigObjectSource` that loads configs from both sources and uses `cs` as a
+   *         fallback for this source
+   */
+  def withFallback(cs: ConfigObjectSource): ConfigObjectSource =
+    new ConfigObjectSource(Result.zipWith(config, cs.value().right.map(_.toConfig))(_.withFallback(_)))
+
+  /**
+   * Returns a `ConfigObjectSource` that provides the same config as this one, but falls back to
+   * providing an empty config when the source cannot be read. It can be used together with
+   * `.withFallback` to specify optional config files to be merged (like `reference.conf`).
+   *
+   * @return a new `ConfigObjectSource` that provides the same config as this one, but falls back
+   *         to an empty config if it cannot be read.
+   */
+  def optional: ConfigObjectSource =
+    recoverWith { case ConfigReaderFailures(_: CannotRead, Nil) => Right(ConfigFactory.empty) }
+
+  /**
+   * Applies a function `f` if this source returns a failure, returning an alternative config in
+   * those cases.
+   *
+   * @param f the function to apply if this source returns a failure
+   * @return a new `ConfigObjectSource` that provides an alternative config in case this source
+   *         fails
+   */
+  def recoverWith(f: PartialFunction[ConfigReaderFailures, Result[Config]]): ConfigObjectSource =
+    new ConfigObjectSource(config.left.flatMap(f))
+}
+
+object ConfigObjectSource {
+
+  /**
+   * Creates a `ConfigObjectSource` from a `Result[Config]`. The provided argument is allowed
+   * to change value over time.
+   *
+   * @param conf the config to be provided by this source
+   * @return a `ConfigObjectSource` providing the given config.
+   */
+  def apply(conf: => Result[Config]): ConfigObjectSource =
+    new ConfigObjectSource(conf)
+
+  /**
+   * Creates a `ConfigObjectSource` from a `ConfigObjectCursor`.
+   *
+   * @param cur the cursor to be provided by this source
+   * @return a `ConfigObjectSource` providing the given cursor.
+   */
+  def fromCursor(cur: ConfigObjectCursor): ConfigObjectSource =
+    new ConfigObjectSource(Right(cur.value.toConfig))
+
+  /**
+   * Creates a `ConfigObjectSource` from a `FluentConfigCursor`.
+   *
+   * @param cur the cursor to be provided by this source
+   * @return a `ConfigObjectSource` providing the given cursor.
+   */
+  def fromCursor(cur: FluentConfigCursor): ConfigObjectSource =
+    new ConfigObjectSource(cur.asObjectCursor.right.map(_.value.toConfig))
+}
+
+object ConfigSource {
```

Comment:
It seems like this object's methods are almost entirely tied to typesafe config. Do you think it's worth trying to rename or partition or comment it in such a way that it leaves semantic room for other sources of config?

Reply:
I can add some Scaladoc to explain this 👍 It makes sense that we have a lot of Typesafe Config specific methods here since it's part of the HOCON spec and useful for migration, but I intend to add a `YamlConfigSource` later in pureconfig-yaml. One awesome thing about this refactoring is that people using yaml will be able to mix sources, including `reference.conf`, into their YAML application files.

Reply:
Done in 9e82848.